File: python/src/azoth/process/reference/component_splitter.py
"""``process.component_splitter`` - a per-component routing, with each outlet flashed.

Spec: ``specs/models/process/component_splitter.toml``

The Python twin of ``crates/azoth-process/src/models/component_splitter.rs`` over
``crates/azoth-process/src/kernels/component_splitter.rs``, written to mirror them.

# The factor is per component and the outlet count is two

``ComponentSplitter.run`` loops ``for i in 0..2`` and reads ``splitFactor[k]`` as the
fraction of component *k* routed to the overhead, so the outlets carry different
*compositions* - which is what separates this from ``process.splitter``.

# Each outlet is flashed, and the class's enthalpies are not

``run`` sets the component moles on an empty fluid, calls ``init(0)`` and runs a ``TPflash``,
so each outlet is a state of its own. The class then *reports* enthalpies that are not those
states: a fresh NeqSim fluid at the first row's bottoms gives ``-20309.24832914077`` where the
class gives ``-14431.31737097011``. This computes the state's, as ``process.splitter`` does
for the same defect.
"""

from __future__ import annotations

from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import ComponentSplitterResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.process.reference.splitter import _route as split_route


def component_splitter(
    components: list[str],
    feed_n: Q,
    feed_z: list[float],
    feed_p: Q,
    feed_t: Q,
    split_factors: list[float],
) -> ComponentSplitterResult:
    """Divide a stream between two outlets component by component.

    Args:
        components: the fluid's substances, by name.
        feed_n: the inlet's molar flow.
        feed_z: the inlet composition.
        feed_p: the inlet pressure, which both outlets flash at.
        feed_t: the inlet temperature, which both outlets flash at.
        split_factors: the fraction of each component to the overhead, one per component.

    Returns:
        Both outlets' records.

    Raises:
        InvalidInputError: where the factors are not one per component, or any is outside
            ``[0, 1]``.

    Example:
        >>> import azoth
        >>> q = azoth.ureg.Quantity
        >>> r = component_splitter(
        ...     ["methane", "n-butane", "n-pentane"],
        ...     q(1.0, "mol/s"),
        ...     [0.5, 0.3, 0.2],
        ...     q(20.0, "bar"),
        ...     q(300.0, "K"),
        ...     [0.98, 0.05, 0.02],
        ... )
        >>> round(r.overhead_n.to("mol/s").magnitude, 3)
        0.509
    """
    spec = _spec()
    checks = checks_for(spec)
    warnings: list[Warning] = []

    n = input_to_si(spec, "feed_n", feed_n)
    p = input_to_si(spec, "feed_p", feed_p)
    t = input_to_si(spec, "feed_t", feed_t)

    apply_checks(
        checks.on_input,
        {"feed_t": t, "split_factors": split_factors[0] if split_factors else None}.get,
        warnings,
    )

    if len(split_factors) != len(feed_z):
        raise InvalidInputError(
            "split_factors",
            f"a component split routes each component, so the factors are one per component: "
            f"{len(split_factors)} factors for {len(feed_z)} components",
        )
    for index, factor in enumerate(split_factors):
        if not 0.0 <= factor <= 1.0:
            raise InvalidInputError(
                "split_factors",
                f"a component's split fraction is a fraction, and component {index}'s is {factor}",
            )

    def outlet(amounts: list[float]) -> tuple[float, list[float]]:
        total = sum(amounts)
        if total <= 0.0:
            return 0.0, list(feed_z)
        return total, [amount / total for amount in amounts]

    overhead_amounts = [n * zi * f for zi, f in zip(feed_z, split_factors, strict=True)]
    bottoms_amounts = [n * zi * (1.0 - f) for zi, f in zip(feed_z, split_factors, strict=True)]
    overhead_n, overhead_z = outlet(overhead_amounts)
    bottoms_n, bottoms_z = outlet(bottoms_amounts)

    # Each outlet is flashed at the feed's own temperature and pressure.
    overhead = split_route(components, t, p, overhead_z, [1.0])
    bottoms = split_route(components, t, p, bottoms_z, [1.0])

    return ComponentSplitterResult(
        overhead_n=from_si(overhead_n, "mol/s"),
        overhead_z=tuple(overhead_z),
        overhead_p=from_si(p, "Pa"),
        overhead_t=from_si(t, "K"),
        overhead_h=from_si(overhead.h_in, "J/mol"),
        bottoms_n=from_si(bottoms_n, "mol/s"),
        bottoms_z=tuple(bottoms_z),
        bottoms_p=from_si(p, "Pa"),
        bottoms_t=from_si(t, "K"),
        bottoms_h=from_si(bottoms.h_in, "J/mol"),
        warnings=tuple(warnings),
    )


def _spec() -> dict[str, object]:
    """The generated spec, imported here so the module has no import-time cycle."""
    from azoth._models_gen import model

    return model("process.component_splitter")

File: crates/azoth-process/src/models/component_splitter.rs
//! `process.component_splitter` - the component splitter's kernel as a registered id.
//!
//! Spec: `specs/models/process/component_splitter.toml`. The arithmetic is
//! [`crate::kernels::component_splitter`]; what is here is the boundary a case and a
//! cross-impl test address.
//!
//! **Two named outlets rather than a `many` one**, because the class fixes the count at two
//! and the palette entry's `products` is corrected to match in the commit that carries this
//! kernel.

use azoth_core::units::{MolarEnergy, Pressure, ThermodynamicTemperature, joules_per_mole};
use azoth_core::{CalcResult, Result, Warning, apply_checks};

use crate::kernels::component_splitter as kernel;
use crate::model_gen;
use crate::stream::Stream;

/// Result of `process.component_splitter`.
#[derive(Debug, Clone, PartialEq)]
pub struct ComponentSplitterResult {
    /// Overhead molar flow, mol/s.
    pub overhead_n: f64,
    /// Overhead composition.
    pub overhead_z: Vec<f64>,
    /// Overhead pressure.
    pub overhead_p: Pressure,
    /// Overhead temperature.
    pub overhead_t: ThermodynamicTemperature,
    /// Overhead molar enthalpy, the state's at its own composition.
    pub overhead_h: MolarEnergy,
    /// Bottoms molar flow, mol/s.
    pub bottoms_n: f64,
    /// Bottoms composition.
    pub bottoms_z: Vec<f64>,
    /// Bottoms pressure.
    pub bottoms_p: Pressure,
    /// Bottoms temperature.
    pub bottoms_t: ThermodynamicTemperature,
    /// Bottoms molar enthalpy, the state's at its own composition.
    pub bottoms_h: MolarEnergy,
    /// Caveats.
    pub warnings: Vec<Warning>,
}

impl CalcResult for ComponentSplitterResult {
    const CALC_ID: &'static str = "process.component_splitter";
    const FIELDS: &'static [&'static str] = &[
        "overhead_n",
        "overhead_z",
        "overhead_p",
        "overhead_t",
        "overhead_h",
        "bottoms_n",
        "bottoms_z",
        "bottoms_p",
        "bottoms_t",
        "bottoms_h",
        "warnings",
    ];

    fn warnings(&self) -> &[Warning] {
        &self.warnings
    }
}

/// Divide a stream between two outlets component by component.
///
/// # Errors
/// [`azoth_core::AzothError::InvalidInput`] if the factors are not one per component, or if
/// any is outside `[0, 1]`, and whatever the databank or a flash refuses.
pub fn component_splitter(
    components: &[String],
    feed_n: f64,
    feed_z: &[f64],
    feed_p: Pressure,
    feed_t: ThermodynamicTemperature,
    split_factors: &[f64],
) -> Result<ComponentSplitterResult> {
    let spec = &model_gen::COMPONENT_SPLITTER_SPEC;
    let mut warnings = Vec::new();
    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "feed_t" => Some(feed_t.value),
            "split_factors" => split_factors.first().copied(),
            _ => None,
        },
        &mut warnings,
    )?;

    let feed = Stream::from_pt(components.to_vec(), feed_z.to_vec(), feed_n, feed_p, feed_t)?;
    let (overhead, bottoms) = kernel(&feed, split_factors)?;

    Ok(ComponentSplitterResult {
        overhead_n: overhead.n,
        overhead_z: overhead.z,
        overhead_p: overhead.p,
        overhead_t: overhead.t,
        overhead_h: joules_per_mole(overhead.h.value),
        bottoms_n: bottoms.n,
        bottoms_z: bottoms.z,
        bottoms_p: bottoms.p,
        bottoms_t: bottoms.t,
        bottoms_h: joules_per_mole(bottoms.h.value),
        warnings,
    })
}
